### PR TITLE
SAMZA-2615: Metrics for Application Master High Availability

### PR DESCRIPTION
**Pull request by mynameborat**
Description:
Application Master High Availability feature was introduced as part of PRs: #1442, #1448, #1449 and #1450. We add the relevant metrics for the components introduced as part of AM HA.

Changes:
- Metrics for Dynamic Container Heartbeat mechanism
- Refactored ContainerHeartbeatMonitor to take config object
- Metrics clean up for metadata management
- Metrics documentation update
- Additional metrics for Application Master to track the number of containers carried from previous attempt

Tests:
- Updated the unit tests for container heartbeat monitor
- Updated the tests for job coordinator metadata manager
- Updated YarnResourceClusterManager tests

API Changes: None
Upgrade Instructions: None
Usage Instructions: The metrics should be available by default for jobs that have AM HA enabled

### CODE REVIEW

File: docs/learn/documentation/versioned/container/metrics-table.html [hunk 1]
```diff
@@ -977,6 +983,57 @@ <h1>Samza Metrics Reference</h1>
...
+    <tr>
+        <th colspan="2" class="section" id="job-coordinator-metadata-manager-metrics">org.apache.samza.coordinator.JobCoordinatorMetadataManager.JobCoordinatorMetadataManagerMetrics<br><span style="font-weight: normal;margin-left:40px;"><b>Note</b>: The following metrics are applicable when Application Master High Availability is enabled</span></th>
+    </tr>
+    <tr>
+        <td>application-attempt-count</td>
+        <td>Denotes the application attempt count within the scope of a single deployment</td>
+    </tr>
+    <tr>
+        <td>config-changed</td>
+        <td>Denotes configuration changed across attempts within the scope of a single deployment</td>
```
Comment (PawasChhokra) [inline docs/learn/documentation/versioned/container/metrics-table.html:995]: Shouldn't this be - "Denotes the number of times the configuration changed..."

File: docs/learn/documentation/versioned/container/metrics-table.html [hunk 2]
```diff
@@ -977,6 +983,57 @@ <h1>Samza Metrics Reference</h1>
...
+    </tr>
+    <tr>
+        <td>job-model-changed</td>
```
Comment (PawasChhokra) [inline docs/learn/documentation/versioned/container/metrics-table.html:998]: Same as above - "Denotes the number of times the job model changed..."

File: docs/learn/documentation/versioned/container/metrics-table.html [hunk 3]
```diff
@@ -977,6 +983,57 @@ <h1>Samza Metrics Reference</h1>
...
+    <tr>
+        <th colspan="2" class="section" id="container-heartbeat-monitor-metrics">org.apache.samza.container.ContainerHeartbeatMonitor.ContainerHeartbeatMetrics<br><span style="font-weight: normal;margin-left:40px;"><b>Note</b>: The following metrics are applicable when Application Master High Availability is enabled</span></th>
+    </tr>
+    <tr>
+        <td>heartbeat-discovery-time-ms</td>
+        <td>Time taken in millis for the container to discover new AM in the event AM changes for heartbeat establishment</td>
+    </tr>
+    <tr>
+        <td>heartbeat-established-failure-count</td>
+        <td>Number of times failed to establish heartbeat in the event of AM changes</td>
```
Comment (PawasChhokra) [inline docs/learn/documentation/versioned/container/metrics-table.html:1027]: Maybe reword it as - "Number of failed attempts to establish heartbeat in the event of AM changes"

File: samza-core/src/main/java/org/apache/samza/container/ContainerHeartbeatMonitor.java [hunk 4]
```diff
@@ -170,4 +212,55 @@ public Thread newThread(Runnable runnable) {
...
   }
+
+  static final class ContainerHeartbeatMetrics {
+    private static final String GROUP = "ContainerHeartbeatMonitor";
+    private static final String HEARTBEAT_DISCOVERY_TIME_MS = "heartbeat-discovery-time-ms";
+    private static final String HEARTBEAT_ESTABLISHED_FAILURE_COUNT = "heartbeat-established-failure-count";
+    private static final String HEARTBEAT_ESTABLISHED_WITH_NEW_AM_COUNT = "heartbeat-established-with-new-am-count";
+    private static final String HEARTBEAT_EXPIRED_COUNT = "heartbeat-expired-count";
+
+    private final Counter heartbeatEstablishFailureCount;
```
Comment (PawasChhokra) [inline samza-core/src/main/java/org/apache/samza/container/ContainerHeartbeatMonitor.java:223]: nit - Change `heartbeatEstablishFailureCount` to `heartbeatEstablishedFailureCount`

**Review by PawasChhokra — COMMENTED** [review]
Left a few comments about metric docs.

**Review by lakshmi-manasa-g — COMMENTED** [review]
finished a pass

File: docs/learn/documentation/versioned/container/metrics-table.html [hunk 5]
```diff
@@ -977,6 +983,57 @@ <h1>Samza Metrics Reference</h1>
...
+        <td>job-model-changed</td>
+        <td>Denotes job model changed across attempts within the scope of a single deployment</td>
+    </tr>
+    <tr>
+        <td>metadata-read-failed-count</td>
+        <td>Read failure count from the underlying metadata store</td>
+    </tr>
+    <tr>
+        <td>metadata-write-failed-count</td>
+        <td>Write failure count to the underlying metadata store</td>
```
[the remark this answers is not in the document]
Comment (mynameborat) [reply docs/learn/documentation/versioned/container/metrics-table.html:1007]: It does write to underlying metadata store. the fact these apply for AM HA is called out in the metrics documentation in the header.

File: samza-core/src/main/java/org/apache/samza/container/ContainerHeartbeatMonitor.java [hunk 6]
```diff
@@ -91,6 +113,7 @@ public void start() {
     scheduler.scheduleAtFixedRate(() -> {
       ContainerHeartbeatResponse response = containerHeartbeatClient.requestHeartbeat();
       if (!response.isAlive()) {
+        metrics.incrementHeartbeatExpiredCount();
```
[the remark this answers is not in the document]
Comment (mynameborat) [reply samza-core/src/main/java/org/apache/samza/container/ContainerHeartbeatMonitor.java:116]: yeah. there is no concept of re-establishing the heartbeat prior to AM HA. but this metric is still useful regardless.

File: samza-core/src/main/java/org/apache/samza/container/ContainerHeartbeatMonitor.java [hunk 7]
```diff
@@ -150,8 +183,17 @@ ContainerHeartbeatClient createContainerHeartbeatClient(String coordinatorUrl, S
...
 
+  @VisibleForTesting
+  ContainerHeartbeatMetrics getMetrics() {
+    return metrics;
+  }
+
   private void forceExit(String message, int timeout) {
     scheduler.schedule(() -> {
+      if (started) {
+        reporters.values().forEach(MetricsReporter::stop);
```
[the remark this answers is not in the document]
Comment (mynameborat) [reply samza-core/src/main/java/org/apache/samza/container/ContainerHeartbeatMonitor.java:194]: `onContainerExpired.run` will invoke the stop on the `ContainerHeartbeatMonitor` which should stop all the reporters.

File: samza-core/src/main/java/org/apache/samza/container/ContainerHeartbeatMonitor.java [hunk 8]
```diff
@@ -100,7 +123,7 @@ public void start() {
           } catch (Exception e) {
             // On exception in re-establish connection with new AM, force exit.
             LOG.error("Exception trying to connect with new AM", e);
-            forceExit("failure in establishing cconnection with new AM", 0);
+            forceExit("failure in establishing cconnection with new AM", SHUTDOWN_TIMOUT_MS);
```
[the remark this answers is not in the document]
Comment (mynameborat) [reply samza-core/src/main/java/org/apache/samza/container/ContainerHeartbeatMonitor.java:126]: I originally didn't intend to do this and hence the tests don't assert this. reflecting back, I ended up adding here and also including them in tests.

File: docs/learn/documentation/versioned/container/metrics-table.html [hunk 9]
```diff
@@ -977,6 +983,57 @@ <h1>Samza Metrics Reference</h1>
...
+        <td>heartbeat-established-failure-count</td>
+        <td>Number of times failed to establish heartbeat in the event of AM changes</td>
+    </tr>
+    <tr>
+        <td>heartbeat-established-with-new-am-count</td>
+        <td>Number of times heartbeat is established with the new AM in the event of AM changes</td>
+    </tr>
+    <tr>
+        <td>heartbeat-expired-count</td>
+        <td>Number of times heartbeat expired with the active AM</td>
```
[the remark this answers is not in the document]
Comment (mynameborat) [reply docs/learn/documentation/versioned/container/metrics-table.html]: these are container heartbeat metrics and hence container level metrics.

File: docs/learn/documentation/versioned/container/metrics-table.html [hunk 10]
```diff
@@ -977,6 +983,57 @@ <h1>Samza Metrics Reference</h1>
...
+        <td>new-deployment</td>
+        <td>Denotes a new deployment due to changes in metadata either attributed to config or job model</td>
+    </tr>
+    </tbody>
+    <tr>
+        <th colspan="2" class="section" id="container-heartbeat-monitor-metrics">org.apache.samza.container.ContainerHeartbeatMonitor.ContainerHeartbeatMetrics<br><span style="font-weight: normal;margin-left:40px;"><b>Note</b>: The following metrics are applicable when Application Master High Availability is enabled</span></th>
+    </tr>
+    <tr>
+        <td>heartbeat-discovery-time-ms</td>
+        <td>Time taken in millis for the container to discover new AM in the event AM changes for heartbeat establishment</td>
```
[the remark this answers is not in the document]
Comment (mynameborat) [reply docs/learn/documentation/versioned/container/metrics-table.html:1023]: hence I referred as discovery as opposed to success or failure.

**Review by PawasChhokra — APPROVED** [review]

**Review by lakshmi-manasa-g — APPROVED** [review]
lgtm
